iree/compiler/Dialect/LinalgExt/Transforms/Tiling.cpp

```diff
@@ -131,10 +131,6 @@ static FailureOr<TiledOp> tileInterfaceOpImpl(
   // Generate an scf.for for the current loop depth.
   Value lb = loopBounds[loopDepth].offset;
   Value ub = loopBounds[loopDepth].size;
-  if (!matchPattern(loopBounds[loopDepth].stride, m_One())) {
```

Comment:
an you give some more information as to why this is removed. I am not getting the full context here.

Reply:
It is a really good question. The short answer is that the loop stride is not 1 in all of the FFT ops. They are `2,4,8,16,...`. It is tilable if the tileSizes is a multiple of stride.

Let's take 1D FFT as an example and let me put some of my thoughts below.

In DispatchLinalgOnTensors pass, we will set the tile size to `workgroup_size * workgroup_count`.

```mlir
    %workgroup_size_0 = flow.dispatch.workgroup.size[0] : index
    %workgroup_id_0 = flow.dispatch.workgroup.id[0] : index
    %workgroup_count_0 = flow.dispatch.workgroup.count[0] : index
    %18 = affine.apply affine_map<(d0)[s0] -> (d0 * s0)>(%workgroup_id_0)[%workgroup_size_0]
    %19 = affine.apply affine_map<(d0)[s0] -> (d0 * s0)>(%workgroup_count_0)[%workgroup_size_0]
    scf.for %arg4 = %18 to %c1024_22 step %19 {
      %20 = affine.min affine_map<(d0)[s0] -> (s0, -d0 + 1024)>(%arg4)[%workgroup_size_0]
      %21 = flow.dispatch.tensor.load %arg2, offsets = [%arg4], sizes = [%20], strides = [1] : !flow.dispatch.tensor<readwrite:1024xf32> -> tensor<?xf32>
      %22 = flow.dispatch.tensor.load %arg3, offsets = [%arg4], sizes = [%20], strides = [1] : !flow.dispatch.tensor<readwrite:1024xf32> -> tensor<?xf32>
      %23:2 = linalg_ext.fft {__internal_linalg_transform__ = "workgroup"} ins(%c2, %16, %17 : index, tensor<2xf32>, tensor<2xf32>) outs(%21, %22 : tensor<?xf32>, tensor<?xf32>) : tensor<?xf32>, tensor<?xf32>
      flow.dispatch.tensor.store %23#0, %arg2, offsets = [%arg4], sizes = [%20], strides = [1] : tensor<?xf32> -> !flow.dispatch.tensor<readwrite:1024xf32>
      flow.dispatch.tensor.store %23#1, %arg3, offsets = [%arg4], sizes = [%20], strides = [1] : tensor<?xf32> -> !flow.dispatch.tensor<readwrite:1024xf32>
    }
    flow.return
  }
```

When I look into this, my first question here is what does tileSize mean. Is it the total size of workload that a launch should handle? Or it is the number of workload should be handled by the thread?

Take M=2 as an example. If a thread is expected to take 10xf32 subtensor and generate a 10xf32 subtensor. Should the tileSize be `5` or `10`? 

In FFT op, the stride of the loop is a power of 2 (i.e. M). It is tilable if a launch handle a multiple of M sizes. There is not a hard limit on the stride.

I prototyped two approaches.

1. Assume tileSizes is the workload size of a launch (which fits current situation more). This requires the tileSizes to be a multiple of the stride. In this case, we configure the kernels in KernelConfig.cpp/KernelDispatch.cpp.

2. Assume tileSize is the number of workload of a launch, (say that there are `tileSize` of blocks that need to be handled). In this case, the step would be `loop.stride * tileSize`, which naturally meets the requirement -- being a multiple of stride.

In either case, we will have to configure it in KernelConfig.cpp/KernelDispatch.cpp. And the stride can be non-one for both cases.

I ended up with approach 1 because it meets the naming more.

Reply:
THanks for that. So let me also dump some thoughts on this.

1) Tile size means the amount of workload per workgroup. This means the `workgroup_size` that is used at the flow level is a symbolic representation of `m * coarsening_factor`. The `coarsening_factor` is essentially how many `m`s will each workgroup perform. The heuristic you want is as `m` gets smaller `coarseing_factor` gets larger.

2) I think it comes down to change in the number of loops the fft op is exposing. Right now it is exposing as many loops as the rank of the output. As we discussed, it should probably be rank + 1 (I think you were thinking of this initially with fft ops, and I suggested `rank` which I am revisiting now). Apart from the "batch" loops, the fft itself has to loops, one over "m" subview and other over elements within the "m". Both of these loops could have stride of 1. Actually going through the exercise of writing a tiling implementation makes this apparent cause without that we break some invariants of the tiling algorithm (which is fine to remove for now, but we should put it back).

Reply:
Thanks for this! I put some more ideas here and plan to land it as it is. I will file a Github issue and put the issue link to the comment, so people would know that we do want `stride==1` constraint. 

=====

I think the `rank+1` definition will help the point one. We will be able to do second tiling on maybe `coarsening_factor`.

However, I'd like to point out that we will hit the same issue. The for-loops in FFT is

```cpp
int m = 1 << s;
int mh = m >> 1;
for (int k = 0; k < n; k += m) {
  for (int j = 0; j < mh; ++j) {
     cplx w = omega[s][j];
      cplx t = w * a[k + j + mh];
      cplx u = a[k + j];

      a[k + j] = u + t;
      a[k + j + mh] = u - t;
  }
}
```

The stride of the first loop is `m` in this case. We can probably look it in another way:

```cpp
int m = 1 << s;
int mh = m >> 1;
for (int new_k = 0; new_k < n / m; ++new_k) {
  for (int j = 0; j < mh; ++j) {
     int k = new_k * m;
     cplx w = omega[s][j];
     cplx t = w * a[k + j + mh];
     cplx u = a[k + j];

     a[k + j] = u + t;
     a[k + j + mh] = u - t;
  }
}
```

With this form, we need more exploration to see how it fit into Linalg rationale. I roughly look into convert-to-loops patterns, and think it might be a workable solution.

Current lowering:

https://github.com/google/iree/blob/5a2ff279e41c292e5bf54c1e6275849c7fb730a4/iree/compiler/Dialect/LinalgExt/Transforms/test/convert_to_loops.mlir#L348-L363

With new form, we can simply update indices of subview to get the correct form.

E.g.,

```mlir
%[[L_REAL_SLICE:.+]] = memref.subview %[[REAL]][%[[K]] * M] [%[[HM]]] [1] 
```

Note that tiling is working in this case, since this is just "how to expand a fft op to loops/stds".

Reply:
Thats exactly what I had in mind as well!